numba/cuda/decorators.py

```diff
@@ -107,14 +112,20 @@ def _jit(func):
             if cache:
                 disp.enable_caching()
 
-            if device:
-                from numba.core import typeinfer
-                with typeinfer.register_dispatcher(disp):
-                    disp.compile_device(argtypes)
-            else:
-                disp.compile(argtypes)
+            for sig in signatures:
+                argtypes, restype = sigutils.normalize_signature(sig)
+
+                if restype and not device and restype != types.void:
+                    raise TypeError("CUDA kernel must have void return type.")
+
+                if device:
+                    from numba.core import typeinfer
+                    with typeinfer.register_dispatcher(disp):
+                        disp.compile_device(argtypes)
+                else:
+                    disp.compile(argtypes)
```

Comment:
Is it worth splitting this loop to:
1. iterate through the signatures, normalizing them, checking they are compliant and then storing the normalized result.
2. iterate through the stored normalized results invoking compilation on each.

This to create a fast path for ensuring signatures are compliant before starting to compile them.

Reply:
I'd rather keep the loop as it is for simplicity - if the decorator is going to raise an exception, I'm not sure we need to optimize performance for that case.

Reply:
Thanks for considering it, I agree, let's not optimise until there's a proven need.